engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java

```diff
@@ -1079,16 +1109,179 @@ public int hashCode()
+
+    public ModuleMeta(String name, Module module, DefaultAttributeMap attributeMap)
+    {
+      LOG.debug("Initializing {} as {}", name, module.getClass().getName());
+      this.name = name;
+      this.module = module;
+      this.id = logicalOperatorSequencer.decrementAndGet();
+      this.attributes = attributeMap;
+      this.dag = new LogicalPlan();
+    }
+
+    @Override
+    public String getName()
+    {
+      return name;
+    }
+
+    @Override
+    public Module getModule()
+    {
+      return module;
+    }
+
+    @Override
+    public DAG.InputPortMeta getMeta(InputPort<?> port)
+    {
+      return null;
+    }
+
+    @Override
+    public DAG.OutputPortMeta getMeta(OutputPort<?> port)
+    {
+      return null;
+    }
+
+    @Override
+    public Attribute.AttributeMap getAttributes()
+    {
+      return null;
+    }
+
+    @Override
+    public <T> T getValue(Attribute<T> key)
+    {
+      return null;
+    }
+
+    @Override
+    public void setCounters(Object counters)
+    {
+
+    }
+
+    @Override
+    public void sendMetrics(Collection<String> metricNames)
+    {
+
+    }
+
+    public String getParentModuleName()
+    {
+      return parentModuleName;
+    }
+
+    public void setParentModuleName(String parentModuleName)
+    {
+      this.parentModuleName = parentModuleName;
+      if (parentModuleName != null)
+        this.name = this.parentModuleName + LogicalPlan.MODULE_NAMESPACE_SEPARATOR + this.name;
+    }
+
+    public LinkedHashMap<InputPortMeta, StreamMeta> getInputStreams()
+    {
+      return inputStreams;
+    }
+
+    public LinkedHashMap<OutputPortMeta, StreamMeta> getOutputStreams()
+    {
+      return outputStreams;
+    }
+
+    public int getId()
+    {
+      return id;
+    }
+
+    public Integer getNindex()
+    {
+      return nindex;
+    }
+
+    public Integer getLowlink()
+    {
+      return lowlink;
+    }
+
+    public LogicalPlan getDag()
+    {
+      return dag;
+    }
+
+    private void writeObject(ObjectOutputStream out) throws IOException
+    {
+      //getValue2(OperatorContext.STORAGE_AGENT).save(operator, id, Checkpoint.STATELESS_CHECKPOINT_WINDOW_ID);
+      out.defaultWriteObject();
+      FSStorageAgent.store(out, module);
+    }
+
+    private void readObject(ObjectInputStream input) throws IOException, ClassNotFoundException
+    {
+      input.defaultReadObject();
+      // TODO: not working because - we don't have the storage agent in parent attribuet map
+      //operator = (Operator)getValue2(OperatorContext.STORAGE_AGENT).load(id, Checkpoint.STATELESS_CHECKPOINT_WINDOW_ID);
+      module = (Module)FSStorageAgent.retrieve(input);
+    }
+
+    public void flattenModule(LogicalPlan parentDAG, Configuration conf)
+    {
+      module.populateDAG(dag, conf);
+      for (ModuleMeta subModuleMeta : dag.getAllModules()) {
+        subModuleMeta.flattenModule(dag, conf);
+        subModuleMeta.setParentModuleName(this.name);
+      }
+      dag.applyStreamLinks();
```

Comment:
The calling function flattenDAG in LogicalPlanConfiguration is also calling dag.applyStreamLinks(); Why twice?